service/history/historyCache_test.go

```diff
@@ -76,37 +88,72 @@ func (s *historyCacheSuite) SetupTest() {
 	s.cache = newHistoryCache(s.mockShard, s.logger)
 }
 
+func (s *historyCacheSuite) TearDownTest() {
+	s.mockExecutionMgr.AssertExpectations(s.T())
+}
+
 func (s *historyCacheSuite) TestHistoryCachePinning() {
 	s.mockShard.GetConfig().HistoryCacheMaxSize = 2
-	domain := "test_domain"
+	domainID := "test_domain_id"
 	s.cache = newHistoryCache(s.mockShard, s.logger)
 	we := workflow.WorkflowExecution{
-		WorkflowId: common.StringPtr("wf-cache-test"),
+		WorkflowId: common.StringPtr("wf-cache-test-pinning"),
 		RunId:      common.StringPtr(uuid.New()),
 	}
 
-	context, release, err := s.cache.getOrCreateWorkflowExecution(domain, we)
+	context, release, err := s.cache.getOrCreateWorkflowExecution(domainID, we)
 	s.Nil(err)
 
 	we2 := workflow.WorkflowExecution{
-		WorkflowId: common.StringPtr("wf-cache-test"),
+		WorkflowId: common.StringPtr("wf-cache-test-pinning"),
 		RunId:      common.StringPtr(uuid.New()),
 	}
 
 	// Cache is full because context is pinned, should get an error now
-	_, _, err2 := s.cache.getOrCreateWorkflowExecution(domain, we2)
+	_, _, err2 := s.cache.getOrCreateWorkflowExecution(domainID, we2)
 	s.NotNil(err2)
 
 	// Now release the context, this should unpin it.
-	release()
+	release(err2)
 
-	_, release2, err3 := s.cache.getOrCreateWorkflowExecution(domain, we2)
+	_, release2, err3 := s.cache.getOrCreateWorkflowExecution(domainID, we2)
 	s.Nil(err3)
-	release2()
+	release2(err3)
 
 	// Old context should be evicted.
-	newContext, release, err4 := s.cache.getOrCreateWorkflowExecution(domain, we)
+	newContext, release, err4 := s.cache.getOrCreateWorkflowExecution(domainID, we)
 	s.Nil(err4)
 	s.False(context == newContext)
-	release()
+	release(err4)
+}
+
+func (s *historyCacheSuite) TestHistoryCacheClear() {
+	s.mockShard.GetConfig().HistoryCacheMaxSize = 20
+	domainID := "test_domain_id"
+	s.cache = newHistoryCache(s.mockShard, s.logger)
+	we := workflow.WorkflowExecution{
+		WorkflowId: common.StringPtr("wf-cache-test-clear"),
+		RunId:      common.StringPtr(uuid.New()),
+	}
+
+	context, release, err := s.cache.getOrCreateWorkflowExecution(domainID, we)
```

Comment:
can you spin up 2 or more go-routines that do the same set of things you do below ? that way, if there is any race condition, there is a good chance of finding it at unit test time with race detector

Reply:
tested in the update